src/handler/sign_tx.c

```diff
@@ -65,29 +65,46 @@ bool verify_change(change_output_info_t *info, tx_output_t output) {
  *
  * The outputs will be on the global context for sign tx (`tx_info`)
  **/
-void read_change_info_v1(buffer_t *cdata) {
+uint16_t read_change_info_v1(buffer_t *cdata) {
     if (!buffer_read_u8(cdata, &G_context.tx_info.change_len)) {
-        THROW(SW_WRONG_DATA_LENGTH);
+        return SW_WRONG_DATA_LENGTH;
     }
 
     if (G_context.tx_info.change_len > TX_MAX_TOKENS + 1) {
         // Some token change is duplicated or we have more tokens than allowed
-        THROW(SW_WRONG_DATA_LENGTH);
+        return SW_WRONG_DATA_LENGTH;
     }
 
+    // TX_MAX_TOKENS + 1 because we allow 1 change for each token + HTR
+    uint8_t known_indexes[TX_MAX_TOKENS + 1] = {0};
+    uint8_t indexes_len = 0;
+
     for (uint8_t i = 0; i < G_context.tx_info.change_len; i++) {
         change_output_info_t *info = &G_context.tx_info.change_info[i];
 
         // 1 byte for change output index
         if (!buffer_read_u8(cdata, &info->index)) {
-            THROW(SW_WRONG_DATA_LENGTH);
+            return SW_WRONG_DATA_LENGTH;
         }
 
+        // check for duplicates
+        for (uint8_t j = 0; j < indexes_len; j++) {
```

Comment:
Since indexes_len can only go up to `TX_MAX_TOKENS + 1` (which is 11) we don't have to enhance this search.